---
title: "<center> **Code supplementary file** <center>"
author: "<center> Tero Sievänen & Tia-Marje Korhonen <center>"
date: "<center> _`r Sys.Date()`_ <center>"
output:
  html_document:
    code_folding: show
    df_print: paged
    theme: yeti
    highlight: tango
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: no
    number_sections: true
  word_document:
    toc: yes
  pdf_document:
    fig_caption: yes
    toc: yes
---

```{r setup, include=FALSE}
library(rmarkdown)
library(tinytex)
library(knitr)
knitr::opts_chunk$set(echo=TRUE, warning=FALSE, message=FALSE)
```

# Introduction

This code supplementary file was used to perform all differential expression analyses between and within the discovery and cancer cohorts.

***

# R-packages used

A variety of R packages was used for this analysis. All packages used are available from the Comprehensive R Archive Network (CRAN), Bioconductor.org, or Github.

***

# Contact information

All the scripts have been written by Tero Sievänen (Uni. of Jyväskylä) and Tia-Marje Korhonen (Uni. of Jyväskylä).

Correspondence to: tero.o.sievanen@jyu.fi

***

# Preprocessing and read mapping

## Preprocessing

[FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) was used for sequence quality control throughout the pipeline. Sequencing adapter removal, trimming and filtering was done with [FASTX-Toolkit](http://hannonlab.cshl.edu/fastx_toolkit/).

## Read mapping

[Bowtie](http://bowtie-bio.sourceforge.net/index.shtml) aligner was used for mapping the preprocessed high-quality reads to human miR-genome derived from [miRBase v.22](http://www.mirbase.org). A cut-off of >=1M reads per sample was applied before bioinformatic analysis. Mean raw read count was 3,761,804 M and 1,082,304 M was the mean c-miR read count across all samples. The mean alignment rate throughout the study population was 63%.

***
# Analysis

## Step 1 - C-miR count data filtering

This script is used to make a filtered c-miR raw count file. Genes with low counts must be removed before DE-analysis, since their biological significance is low or might be false positives.

```{r step 1 - rawCountFiltering}

# Load packages
library(edgeR) # Package for differential gene expression analysis of RNA-seq data

# Import raw counts file
counts <- read.csv("rawCounts.tsv",header=TRUE,sep="\t")

# Import study design file
targets<- read.csv("phenodata.txt",sep="\t",header=TRUE)

# Add column names
colnames(counts) <-targets$Filename

# Filter phenodata based on sample type
Group1  <- which(targets$Type=="LS")
Group2  <- which(targets$Type=="CTRL")
Group3  <- which(targets$Type=="SRME")
ls <- targets[Group1,]
ctrl <- targets[Group2,]
srme <- targets[Group3,]

# Create a digital gene expression list (DGEList) using edgeR
myDGEList <- DGEList(counts = counts)

# Convert DGEList to counts per million (cpm)
cpm <- cpm(myDGEList)

# Keep only the c-miRs with >1 CPM in at least 70% of samples in a subgroup
keepers <- rowSums(cpm>1)>=108  #70% of 155

# Filter the subgroups (LS,CTRL and SRME)
cpm1 <- cpm[,which(colnames(cpm)%in%ls$Sample)]
cpm2 <- cpm[,which(colnames(cpm)%in%ctrl$Sample)]
cpm3 <- cpm[,which(colnames(cpm)%in%srme$Sample)]

keep1 <- rowSums(cpm1>=1) >= 65   #  70% of 94
keep2 <- rowSums(cpm2>=1) >= 25   #  70% of 37
keep3<- rowSums(cpm3>=1) >= 16   #  70% of 24

# Create a new filtered DGEList
myDGEList.filtered <- myDGEList[(keepers|keep1|keep2|keep3),]

# Create a new filtered raw counts matrix
FilteredCounts <- counts[which(rownames(counts)%in%rownames(myDGEList.filtered$counts)),]

# Create a new .txt file of filtered raw miR counts
write.table(FilteredCounts,"FilteredCounts.txt",sep="\t")
print("Step 1 completed!")
```

***

## Step 2 - Sex difference within _path_MMR_ carriers, sporadic rectal cancer patients, and non-LS control group

This script is used to perform DE-analysis between sexes among _path_MMR_ carriers, sporadic rectal cancer patients (SRME) and non-LS control group (CTRL).

```{r step 2 - sexDifference}

# Load packages
library(tidyverse) # Tidyverse is an opinionated collection of R packages designed for data science
library(gt) # Package for making static data tables
library(DESeq2) # Package for differential gene expression analysis of RNA-seq data
library(plotly) # Package for making interactive plots
library(DT) # Package for making interactive tables

# Read in the filtered c-miR counts table from step 1
counts <- read.csv("FilteredCounts.txt",header=TRUE,sep="\t")

# Add sample names (column names) to raw counts table
colnames(counts) <-targets$Filename

# Choose only the healthy path_MMR carriers (!= cancer | future_ca), n = 81
select <- which(targets$Type=="LS" & targets$Healthy_now=="YES")

# AND/ OR 
#select <- which(targets$Type=="SRME") # No sex difference
#select <- which(targets$Type=="CTRL") # No sex difference

# Create a new filtered counts file 
Counts <- counts[,select]

# New phenofile with only the variables of interest
Targets <- targets[select,]

# Setup design matrix for DE-analysis
condition <- as.character(Targets$Sex) # Condition of interest
batch <- as.character(Targets$NGS) # Batch effect
group_levels <- levels(as.factor(condition))
design <- data.frame(condition=as.factor(condition), batch=batch) # DE-design for the analysis
rownames(design) <- colnames(Counts)
dds <- DESeqDataSetFromMatrix(countData=Counts, colData=design, design = ~ batch + condition)

# DESeq2 DE-analysis of the condition of interest, batch effect taken into account
dds <- DESeq(dds)

# Display results
res <- results(dds, alpha=0.05) # Statistical significance at the level p=< 0.05
resOrdered <- res[order(res$padj),] # Order results based on adjusted p-value
summary(res)
resOrdered
mcols(res)$description

# Create a data frame of the ordered results (top 20)
padj.subset <- head(resOrdered, 20) %>%
  as_tibble(rownames = "miR")

# Create a gene table of the subset
#gt(subset)

# Create a interactive gene table of the subset
datatable(padj.subset, 
          extensions = c('KeyTable', "FixedHeader"), 
          caption = 'Table 1: Differentially expressed c-miRs between sexes in path_MMR carriers',
          options = list(keys = TRUE, searchHighlight = TRUE, pageLength = 20, lengthMenu = c("10", "25", "50", "100"))) %>%
  formatRound(columns=c(2:7), digits=3)

# Create a data frame of all results for plotting
res.df <- as_tibble(res, rownames = "miR")

# Create a volcano plot of results
v1 <- ggplot(data=res.df,
  aes(y=-log10(res$padj), x=res$log2FoldChange, text = paste(rownames(Counts)))) +
  xlab("Log2FC") +
ylab("-log10(Padj)") +
  geom_point(size=2) +
  geom_hline(yintercept = -log10(0.05), linetype="longdash", colour="grey", size=1) +
  geom_vline(xintercept = 1, linetype="longdash", colour="#BE684D", size=1) +
  geom_vline(xintercept = -1, linetype="longdash", colour="#2C467A", size=1) +
  labs(title="Volcano plot",
       subtitle = "Men vs women in LS group",
       caption=paste0("produced on ", Sys.time())) +
  theme_bw()

# Create an interactive volcano plot of results
ggplotly(v1)

# We saw that there is a sex difference within LS group, so to confirm that this is an unique finding, we need to perform the same analysis to SRME and CTRL group also

# Run the script again from the beginning but subset the groups from phenodata

print("Step 2 completed! There is a sex difference only in LS.")
```

***

## Step 3 - _Path_MMR_ variants in LS

This script is used to perform DE-analysis between MLH1 and other variants (MSH2, MSH6 and PMS2) among _path_MMR_ carriers.

```{r step 3 - pathVariantLS}

# Setup design matrix for DE-analysis
condition <- as.character(Targets$step_3) # condition of interest (path_MMR variant)
batch <- as.character (Targets$NGS) # Batch effect 
sex <- as.character(Targets$Sex) # Sex as covariate
group_levels <- levels(as.factor(condition))
design <- data.frame(condition=as.factor(condition), batch=batch, sex=sex)
rownames(design) <- colnames(Counts)
dds <- DESeqDataSetFromMatrix(countData=Counts, colData=design, design = ~ batch + sex + condition)

# DESeq2 DE-analysis of the condition of interest, batch effect and sex as covariates
dds <- DESeq(dds)

# Display results
res <- results(dds, alpha=0.05)
resOrdered <- res[order(res$padj),]
summary(res)
resOrdered
mcols(res)$description

# Create a data frame of the ordered results (top 20)
subset <- head(resOrdered, 20) %>%
  as_tibble(rownames = "miR")

# Create a gene table of the subset
gt(subset)
print("Step 3 completed! There was no difference in c-miR expression among path_MMR variants.")
```

***

## Step 4 - Cancer history in _path_MMR_ carriers

This script is used to perform DE-analysis between _path_MMR carriers_ with or without previous cancer(s). 

```{r step 4 - cancerHistoryLS}

# Setup design matrix for DE-analysis
condition <- as.character (Targets$Previous_ca) # Condition of interest (cancer history, 0 = no, 1 = yes)
batch <- as.character (Targets$NGS) # Batch effect
sex <- as.character(Targets$Sex) # Sex as covariate
group_levels <- levels(as.factor(condition))
design <- data.frame(condition=as.factor(condition), batch=batch, sex=sex)
rownames(design) <- colnames(Counts)
dds <- DESeqDataSetFromMatrix(countData=Counts, colData=design, design = ~ batch + sex + condition)

# DESeq2 DE-analysis of the condition of interest, batch effect and sex as covariates
dds <- DESeq(dds)

# Display results
res <- results(dds, alpha=0.05)
resOrdered <- res[order(res$padj),]
summary(res)
resOrdered
mcols(res)$description

# Create a data frame of the ordered results (top 20)
subset <- head(resOrdered, 20) %>%
  as_tibble(rownames = "miR")

# Create a gene table of the subset
gt(subset)
print("Step 4 completed! There was no difference in c-miR expression between path_MMR carriers with or without cancer history.")
```

***

## Step 5 - Healthy _path_MMR_ carriers vs _path_MMR_ carriers with cancer

This script is used to perform DE-analysis between healthy _path_MMR_ carriers (n=81) and _path_MMR_ carriers with cancer (n=13). 

```{r step 5 - healthyLS_caLS}

# Choose conditions of interest (healthy path_MMR carriers, n = 81 and path_MMR carriers with cancer, n = 13)
select <- which(targets$Type=="LS" & targets$Healthy_now=="YES" | targets$Healthy_now=="NO")

# Create a new filtered counts file
Counts <- counts[,select]

# New phenofile with only the variables of interest
Targets <- targets[select,]

# Setup design matrix for DE-analysis
condition <- as.character (Targets$Healthy_now) # Condition of interest
batch <- as.character (Targets$NGS) # Batch effect
sex <- as.character(Targets$Sex) # Sex as covariate
group_levels <- levels(as.factor(condition))
design <- data.frame(condition=as.factor(condition), batch=batch, sex=sex)
rownames(design) <- colnames(Counts)
dds <- DESeqDataSetFromMatrix(countData=Counts, colData=design, design = ~ batch + sex + condition)

# DESeq2 DE-analysis of the condition of interest, batch effect and sex as covariates
dds <- DESeq(dds)

# Display results
res <- results(dds, alpha=0.05)
resOrdered <- res[order(res$padj),]
summary(res)
resOrdered
mcols(res)$description

# Create a data frame of the ordered results (top 20)
subset <- head(resOrdered, 20) %>%
  as_tibble(rownames = "miR")

# Create a gene table of the subset
gt(subset)
print("Step 5 completed! There was no difference in c-miR expression between path_MMR carriers with or without cancer.")
```

***

## Step 6 - Healthy _path_MMR_ carriers vs non-LS control group

This script is used to perform DE-analysis between healthy _path_MMR_ carriers and CTRL group.

```{r step 6 - LSvsCTRL}

# Choose conditions of interest (healthy path_MMR carriers, n = 81 and CTRL group, n = 37)
select <- which(targets$Type=="LS" & targets$Healthy_now=="YES" | targets$Type=="CTRL")

# Create a new filtered counts file
Counts <- counts[,select]

# New phenofile with only the variables of interest
Targets <- targets[select,]

# Setup design matrix for DE-analysis
condition <- as.character (Targets$Type) # Condition of interest
batch <- as.character (Targets$NGS) # Batch effect
sex <- as.character(Targets$Sex) # Sex as covariate
group_levels <- levels(as.factor(condition))
design <- data.frame(condition=as.factor(condition), batch=batch, sex=sex)
rownames(design) <- colnames(Counts)
dds <- DESeqDataSetFromMatrix(countData=Counts, colData=design, design = ~ batch + sex + condition)

# DESeq2 DE-analysis of the condition of interest, batch effect and sex as covariates
dds <- DESeq(dds)

# Display results
res <- results(dds, alpha=0.05)
resOrdered <- res[order(res$padj),]
summary(res)
resOrdered
mcols(res)$description

# Create a data frame of the ordered results (top 39)
padj.subset <- head(resOrdered, 40) %>%
 as_tibble(rownames = "miR")

# Create a interactive gene table of the subset
datatable(padj.subset, 
          extensions = c('KeyTable', "FixedHeader"), 
          caption = 'Table 2: Differentially expressed c-miRs in healthy path_MMR carriers vs non-LS control group',
          options = list(keys = TRUE, searchHighlight = TRUE, pageLength = 20, lengthMenu = c("5", "10", "15", "50"))) %>%
  formatRound(columns=c(2:7), digits=3)

# Create a data frame of all results for plotting
res.df <- as_tibble(res, rownames = "miR")

# Create a volcano plot of results
v2 <- ggplot(data=res.df,
  aes(y=-log10(res$padj), x=res$log2FoldChange, text = paste(rownames(Counts)))) +
  xlab("Log2FC") +
ylab("-log10(Padj)") +
  geom_point(size=2) +
  geom_hline(yintercept = -log10(0.05), linetype="longdash", colour="grey", size=1) +
  geom_vline(xintercept = 1, linetype="longdash", colour="#BE684D", size=1) +
  geom_vline(xintercept = -1, linetype="longdash", colour="#2C467A", size=1) +
  labs(title="Volcano plot",
       subtitle = "LS healthy vs CTRL",
       caption=paste0("produced on ", Sys.time())) +
  theme_bw()

# Create an interactive volcano plot of results
ggplotly(v2)
print("Step 6 completed! There are 15 upregulated and 25 downregulated c-miRs in healthy path_MMR carriers compared to CTRL group.")
```

***

## Step 7 - Sporadic rectal cancer patients vs healthy _path_MMR_ carriers

This script is used to perform DE-analysis between SRME group and healthy _path_MMR_ carriers.

```{r step 7 - LSvsSRME}

# Choose conditions of interest (healthy path_MMR carriers, n = 81 and SRME, n = 24)
select <- which(targets$Type=="LS" & targets$Healthy_now=="YES" | targets$Type=="SRME")

# Create a new filtered counts file
Counts <- counts[,select]

# New phenofile with only the variables of interest
Targets <- targets[select,]

# Setup design matrix for DE-analysis
condition <- as.character (Targets$Type) # Condition of interest
batch <- as.character (Targets$NGS) # Batch effect
sex <- as.character(Targets$Sex) # Sex as covariate
group_levels <- levels(as.factor(condition))
design <- data.frame(condition=as.factor(condition), batch=batch, sex=sex)
rownames(design) <- colnames(Counts)
dds <- DESeqDataSetFromMatrix(countData=Counts, colData=design, design = ~ batch + sex + condition)

# DESeq2 DE-analysis of the condition of interest, batch effect and sex as covariates
dds <- DESeq(dds)

# Display results
res <- results(dds, alpha=0.05)
resOrdered <- res[order(res$padj),]
summary(res)
resOrdered
mcols(res)$description

# Create a data frame of the ordered results (top 20)
subset<- head(resOrdered, 20) %>%
  as_tibble(rownames = "miR")

# Create a gene table of the subset
gt(subset)
print("Step 7 completed! There was no difference in c-miR expression between SRME group and healthy path_MMR carriers.")
```
***

## Step 8 - Sporadic rectal cancer patients vs _path_MMR_ carriers with cancer

This script is used to perform DE-analysis between SRME group and _path_MMR_ carriers with cancer.

```{r step 8 - SRMEvscaLS}

# Choose conditions of interest (path_MMR with cancer, n = 13 and SRME, n = 24)
select <- which(targets$Type=="LS" & targets$Healthy_now=="NO" | targets$Type=="SRME")

# Create a new filtered counts file
Counts <- counts[,select]

# New phenofile with only the variables of interest
Targets <- targets[select,]

# Setup design matrix for DE-analysis
condition <- as.character (Targets$Type) # Condition of interest
batch <- as.character (Targets$NGS) # Batch effect
sex <- as.character(Targets$Sex) # Sex as covariate
group_levels <- levels(as.factor(condition))
design <- data.frame(condition=as.factor(condition), batch=batch, sex=sex)
rownames(design) <- colnames(Counts)
dds <- DESeqDataSetFromMatrix(countData=Counts, colData=design, design = ~ batch + sex + condition)

# DESeq2 DE-analysis of the condition of interest, batch effect and sex as covariates
dds <- DESeq(dds)

# Display results
res <- results(dds, alpha=0.05)
resOrdered <- res[order(res$padj),]
summary(res)
resOrdered
mcols(res)$description

# Create a data frame of the ordered results (top 20)
subset<- head(resOrdered, 20) %>%
  as_tibble(rownames = "miR")

# Create a gene table of the subset
gt(subset)

print("Step 8 completed! There was no difference in c-miR expression between SRME group and path_MMR carriers with cancer.")
```
***

## Step 9 - Sporadic rectal cancer patients vs non-LS control group

This script is used to perform DE-analysis between SRME group and CTRL group.

```{r step 9 - SRMEvsCTRL}

# Choose conditions of interest (SRME, n = 24, CTRL, n = 37)
select <- which(targets$Type=="SRME" | targets$Type=="CTRL")

# Create a new filtered counts file
Counts <- counts[,select]

# New phenofile with only the variables of interest
Targets <- targets[select,]

# Setup design matrix for DE-analysis
condition <- as.character (Targets$Type) # Condition of interest
batch <- as.character (Targets$NGS) # Batch effect
sex <- as.character(Targets$Sex) # Sex as covariate
group_levels <- levels(as.factor(condition))
design <- data.frame(condition=as.factor(condition), batch=batch, sex=sex)
rownames(design) <- colnames(Counts)
dds <- DESeqDataSetFromMatrix(countData=Counts, colData=design, design = ~ batch + sex + condition)

# DESeq2 DE-analysis of the condition of interest, batch effect and sex as covariates
dds <- DESeq(dds)

# Display results
res <- results(dds, alpha=0.05)
resOrdered <- res[order(res$padj),]
summary(res)
resOrdered
mcols(res)$description

# Create a data frame of the ordered results (top 3)
padj.subset<- head(resOrdered, 4) %>%
  as_tibble(rownames = "miR")

# Create a interactive gene table of the subset
datatable(padj.subset, 
          extensions = c('KeyTable', "FixedHeader"), 
          caption = 'Table 3: Differentially expressed c-miRs in sporadic rectal cancer patients vs non-LS control group',
          options = list(keys = TRUE, searchHighlight = TRUE, pageLength = 20, lengthMenu = c("5", "10", "15", "50"))) %>%
  formatRound(columns=c(2:7), digits=3)

# Create a data frame of all results for plotting
res.df <- as_tibble(res, rownames = "miR")

# Create a volcano plot of results
v3 <- ggplot(data=res.df,
             aes(y=-log10(res$padj), x=res$log2FoldChange)) +
  xlab("Log2FC") +
  ylab("-log10(Padj)") +
  geom_point(size=2) +
  geom_hline(yintercept = -log10(0.05), linetype="longdash", colour="grey", size=1) +
  geom_vline(xintercept = 1, linetype="longdash", colour="#BE684D", size=1) +
  geom_vline(xintercept = -1, linetype="longdash", colour="#2C467A", size=1) +
  labs(title="Volcano plot",
       subtitle = "LS vs SRME",
       caption=paste0("produced on ", Sys.time())) +
  theme_bw()

ggplotly(v3)
print("Step 9 completed! There were 4 upregulated c-miRs in SRME group compared to CTRL.")
```

***

## Step 10 - _Path_MMR_ carriers with cancer vs non-LS control group

This script is used to perform DE-analysis between _path_MMR_ carriers with cancer and CTRL group.

```{r step 10 - caLSvsCTRL}

# Choose conditions of interest (Path_MMR carriers with cancer, n = 13 and CTRL, n = 37)
select <- which(targets$Type=="LS" & targets$Healthy_now=="NO" | targets$Type=="CTRL")

# Create a new filtered counts file
Counts <- counts[,select]

# New phenofile with only the variables of interest
Targets <- targets[select,]

# DE-analysis with DESeq2

# Setup design matrix for DE-analysis
condition <- as.character (Targets$Type) # condition of interest to be tested
batch <- as.character (Targets$NGS) # batch effect
sex <- as.character(Targets$Sex) # Sex as a covariate
group_levels <- levels(as.factor(condition))
design <- data.frame(condition=as.factor(condition), batch=batch, sex=sex)
rownames(design) <- colnames(Counts)
dds <- DESeqDataSetFromMatrix(countData=Counts, colData=design, design = ~ batch + sex + condition)

# DESeq2 DE-analysis of the condition of interest, batch effect taken into account, sex as a covariate
dds <- DESeq(dds)

# Display results
res <- results(dds, alpha=0.05)
resOrdered <- res[order(res$padj),]
summary(res)
resOrdered
mcols(res)$description

# Create a data frame of the ordered results (top 20)
subset <- head(resOrdered, 20) %>%
  as_tibble(rownames = "miR")

# Create a gene table of the subset
#names(subset)<- c("miR","Mean", "log2FC", "SE", "Wald", "p-value", "FDR")
gt(subset)
print("Step 10 completed! There was no difference in c-miR expression between path_MMR carriers with cancer and CTRL group.")
```

# Session info

The output from running 'sessionInfo' is shown below and details all packages and version necessary to reproduce the results in this report.

```{r session info}
sessionInfo()
```